lib/l1_watcher/src/upgrade_tx_watcher.rs

```diff
@@ -307,5 +336,18 @@ pub enum UpgradeTxWatcherError {
     #[error("Timestamp exceeds u64: {0}")]
     TimestampExceedsU64(U256),
     #[error("Incorrect protocol version: {0}")]
-    IncorrectProtocolVersion(U256),
+    IncorrectProtocolVersion(#[from] ProtocolSemanticVersionError),
+}
+
+async fn find_l1_block_by_protocol_version(
+    zk_chain: ZkChain<DynProvider>,
+    protocol_version: ProtocolSemanticVersion,
+) -> anyhow::Result<BlockNumber> {
+    let protocol_version = protocol_version.packed()?;
+
+    util::find_l1_block_by_predicate(Arc::new(zk_chain), move |zk, block| async move {
+        let res = zk.get_raw_protocol_version(block.into()).await?;
+        Ok(res == protocol_version)
```

Comment:
You need to do `res >= protocol_version` here to find the first block where `protocol_version` got activated. Or `res > protocol_version` to find first block where next protocol version got activated (or it will return `latest + 1` if there is none). I think latter makes more sense for you

Apologies that the method doesn't have any documentation, I can see how easy it would be to confuse that predicate here does not need to be monotonic.

Reply:
Fixed, thanks. I do think that the first block where `protocol_version` got activated makes more sense, since we can have `setUpgradeTimestamp` at any point after this block for the next (not yet processed) upgrade.